chalk-solve/src/clauses.rs

```diff
@@ -262,8 +261,19 @@ fn program_clauses_that_could_match<I: Interner>(
                         let qwc =
                             exists_qwc.substitute(interner, &[self_ty.clone().cast(interner)]);
 
-                        builder.push_binders(&qwc, |builder, wc| {
-                            builder.push_fact(wc);
+                        builder.push_binders(&qwc, |builder, wc| match &wc {
+                            // For the implemented traits, we need to elaborate super traits and add where clauses from the trait
+                            WhereClause::Implemented(trait_ref) => {
+                                let mut seen_traits = FxHashSet::default();
```

Comment:
Hmm this query would have to return something like `Binders<Vec<Binders<TraitRef>>>` (outer binders for the original trait's parameters, inner binders for `for<'a>` super trait bounds), right?

Reply:
Yes... in rustc, we skip the outer binders, but you're correct that ideally it'd have a `Binders` type. Perhaps -- just for the purposes of this PR -- you could move this logic to a method on `TraitDatum`?

Reply:
Well, this turned out to be more interesting than expected, because if we have `Foo: for<'a> Bar<'a>` and `Bar<'a>: for<'b> Baz<'a, 'b>` we need to collapse the binders.

I've extracted it to a function for now; I'm not sure whether it makes sense to be on `TraitDatum`, but if you want I can move it there.